### DISCUSSION TITLE
Setting limits on the maximum step of manipulated variables

### DISCUSSION THREAD

**Discussion by KhamutskiyDA** [do-mpc/do-mpc#479] [Q&A]
I am solving an MPC problem formulated in state space. I have control actions set by absolute values. I want to put constraints on their maximum step at each step, i.e. -max_step <= u - u_prev <= max_step
All the methods I looked at did not produce the desired result, I have already started doing it this way
<img width="992" alt="image" src="https://github.com/user-attachments/assets/13e5297d-e069-432f-a517-f4659d96e36f">
The problem is that mpc.data are updated when the best step is selected, that is, when MPC performs optimization over the entire horizon to select the optimal step, then this value is fixed and you get a constraint just on the absolute value of the control variable

I made a crutch and entered the previous value as a variable that lags by 1 step. This works perfectly, but the speed has increased a lot
<img width="665" alt="image" src="https://github.com/user-attachments/assets/ee79420b-8072-4438-96b3-29adb155e573">
<img width="964" alt="image" src="https://github.com/user-attachments/assets/5f4ba52b-d5c2-45d9-9d62-e6838682f43f">
How can I make it so that I can get the previous value all the time and can make a maximum step limit?

It might make sense to use nlp_cons, but I don't know how well it would work.
There was an idea to use tvp, but then they should be updated automatically, not set in advance. How can this be done?

**Comment by KhamutskiyDA** [comment]
I've solved that problem. I needed to add the following structure for constraints. This is a very low-level constraint, so you have to be careful.
```
mpc.prepare_nlp()
u_list = mpc.opt_x['_u']
max_step = 1

for i in range(len(u_list)):
    if i == 0:
        bound = mpc.opt_x['_u', i, 0] - mpc.opt_p['_tvp', i, 'U0t']
    else:
        bound = mpc.opt_x['_u', i, 0] - mpc.opt_x['_u', i-1, 0]
    mpc.nlp_cons.append(bound)
    shape = bound.shape
    mpc.nlp_cons_lb.append(-np.array([max_step]*shape[0]).reshape(shape))
    mpc.nlp_cons_ub.append(np.array([max_step]*shape[0]).reshape(shape))
mpc.create_nlp()
```
This code imposes restrictions on all neighboring _u's. It is very important to specify exactly _tvp at the beginning. The point is that if you specify mpc.u0, mpc.u_pred and similar constructions, they will not be updated in the loop and you will get that the first u must always be close to a fixed value, which will give a step that is limited by the absolute value, not by the difference.

For _tvp, you need to throw them through each step of the main loop. I did it in the following way.
MPC:
```
tvp_template_mpc = mpc.get_tvp_template()
tvp_fun_mpc = lambda t_now: tvp_template_mpc
mpc.set_tvp_fun(tvp_fun_mpc)
```
Simulator:
```
simulator = Simulator(model)
simulator.set_param(t_step=1)
tvp_template_sim = simulator.get_tvp_template()
tvp_fun_sim = lambda t_now: tvp_template_sim
simulator.set_tvp_fun(tvp_fun_sim)
simulator.setup()
```
It is crucial to call them by different names!
Main loop
```
tvp_template_mpc = mpc.get_tvp_template()
tvp_template_sim = simulator.get_tvp_template()

model_horizon = 40
for i in range(model_horizon):
    for k in range(n_horizon+1):
        tvp_template_mpc['_tvp', k, 'U0t'] = u0
    tvp_template_sim['U0t'] = u0
    
    u0 = mpc.make_step(simulator.x0)
    simulator.make_step(u0)
```
The code I give here is a SISO example. In MIMO everything is the same. Unfortunately I can't give you the whole code, but it should be enough for your needs

**Reply by jrbull** [reply]
I implemented your solution, very helpful - thanks. Previously I had tried to enforce limits via the rterm. How do you set your rterm when you have the hard constraints on delta u? Secondly, what is the purpose of setting U0t in tvp_template_sim in your example?
